File: algorithms/easy/take_gifts_from_the_richest_pile.cpp
class Solution {
public:
  long long pickGifts(vector<int> &gifts, int k) {
    priority_queue<int> pq(gifts.begin(), gifts.end());
    for (int i = 0; i < k; i++) {
      int n = pq.top();
      pq.pop();
      pq.push(sqrt(n));
    }
    long long ans = 0;
    while (!pq.empty()) {
      ans += pq.top();
      pq.pop();
    }
    return ans;
  }
};

File: algorithms/easy/take_gifts_from_the_richest_pile.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn pick_gifts(gifts: Vec<i32>, k: i32) -> i64 {
        let mut hp = BinaryHeap::new();
        gifts.into_iter().for_each(|n| hp.push(n as i64));
        for _ in 0..k {
            let n = hp.pop().unwrap();
            hp.push((n as f64).sqrt() as i64);
        }
        hp.into_iter().sum::<i64>()
    }
}
